File: rules/evaluate.ts
export interface FaceState {
  is_occupied: boolean;
  built_type: string;
  height_floors: number;
  value?: number;
  population?: number;
}

export interface FaceEvaluation {
  score: number;
}

export interface Face {
  id: number;
  neighbors: number[];
  boundary_polyline: number[][]; // [ [x, y, z], ... ]
  state: FaceState;
  evaluation: FaceEvaluation;
}

export interface TopologyMetadata {
  map_id: string;
  total_faces: number;
}

export interface TopologyData {
  metadata: TopologyMetadata;
  faces: Face[];
}

export interface EvaluationResult {
  faces: Face[];
  government_tax: number;
  developer_profit: number;
  total_population: number;
}

/**
 * Evaluates the urban economy metrics for the given faces.
 * Ported from evaluate.py to frontend rules.
 */
export function evaluateUrbanEconomy(faces: Face[]): EvaluationResult {
  // 1. Create a map for fast lookup of faces by ID
  const faceMap = new Map<number, Face>();
  
  // Clone the faces to prevent modifying the inputs in-place directly
  const clonedFaces: Face[] = faces.map(f => ({
    ...f,
    state: {
      ...f.state,
      value: f.state.value ?? 0,
      population: f.state.population ?? 0,
    },
    evaluation: {
      ...f.evaluation,
      score: f.evaluation?.score ?? 0,
    }
  }));

  for (const face of clonedFaces) {
    faceMap.set(face.id, face);
  }

  // 2. 第一步：计算所有住宅的价值 (Value)
  for (const face of clonedFaces) {
    if (face.state.built_type !== 'residential') {
      continue;
    }

    let value = 40; // 基础价值基准分

    // 扫描邻居
    for (const neighborId of face.neighbors) {
      const neighbor = faceMap.get(neighborId);
      if (neighbor) {
        const neighborType = neighbor.state.built_type;
        if (neighborType === 'park' || neighborType === 'greenway' || neighborType === 'green') {
          value += 25; // 绿地大幅加分
        } else if (neighborType === 'residential') {
          value -= 5;  // 住宅扎堆，产生拥挤负效应
        }
      }
    }

    // 限制在10-100分
    face.state.value = Math.max(10, Math.min(100, value));
  }

  // 3. 第二步：根据正态分布计算入住人口 (Population)
  const mu = 60;       // 人口最喜欢的黄金房价点
  const sigma = 20;    // 敏感度系数
  const P_max = 150;   // 单个格子最大承载人口

  let totalCityPopulation = 0;
  let totalDeveloperProfit = 0;

  for (const face of clonedFaces) {
    if (face.state.built_type === 'residential') {
      const v = face.state.value ?? 0;
      
      // 高斯/正态分布公式
      const pop = P_max * Math.exp(-Math.pow(v - mu, 2) / (2 * Math.pow(sigma, 2)));
      const roundedPop = Math.round(pop);

      face.state.population = roundedPop;
      face.evaluation.score = v; // 把价值作为前端显示的得分

      // 累计数据
      totalCityPopulation += roundedPop;
      totalDeveloperProfit += roundedPop * v;
    } else {
      // Reset non-residential values
      face.state.value = 0;
      face.state.population = 0;
      face.evaluation.score = 0;
    }
  }

  // 4. 第三步：计算政府总税收
  const taxRate = 0.5;
  const totalGovernmentTax = Math.round(totalCityPopulation * taxRate);

  return {
    faces: clonedFaces,
    government_tax: totalGovernmentTax,
    developer_profit: totalDeveloperProfit,
    total_population: totalCityPopulation
  };
}

File: backend/evaluate.py
import math
import copy

def evaluate_urban_economy(faces):
    """
    Evaluates the urban economy metrics for the given faces.
    Ported from rules/evaluate.ts.
    """
    # Clone the faces to prevent modifying the inputs in-place
    cloned_faces = copy.deepcopy(faces)
    
    # 1. Create a map for fast lookup of faces by ID
    face_map = {face['id']: face for face in cloned_faces}
    
    # 2. First pass: Calculate value for all residentials
    for face in cloned_faces:
        state = face.setdefault('state', {})
        state['value'] = state.get('value', 0)
        state['population'] = state.get('population', 0)
        
        evaluation = face.setdefault('evaluation', {})
        evaluation['score'] = evaluation.get('score', 0)
        
        if state.get('built_type') != 'residential':
            continue
            
        value = 40  # Base value baseline score
        
        # Scan neighbors
        for neighbor_id in face.get('neighbors', []):
            neighbor = face_map.get(neighbor_id)
            if neighbor:
                neighbor_type = neighbor.get('state', {}).get('built_type', '')
                if neighbor_type in ('park', 'greenway', 'green'):
                    value += 25  # Large green bonus
                elif neighbor_type == 'residential':
                    value -= 5   # Residential crowding penalty
                    
        state['value'] = max(10, min(100, value))
        
    # 3. Second pass: Calculate population based on normal distribution
    mu = 60       # Golden price point for population
    sigma = 20    # Sensitivity coefficient
    P_max = 150   # Maximum population capacity per grid
    
    total_city_population = 0
    total_developer_profit = 0
    
    for face in cloned_faces:
        state = face.get('state', {})
        evaluation = face.get('evaluation', {})
        
        if state.get('built_type') == 'residential':
            v = state.get('value', 0)
            
            # Gaussian/normal distribution formula
            pop = P_max * math.exp(-((v - mu) ** 2) / (2 * (sigma ** 2)))
            rounded_pop = round(pop)
            
            state['population'] = rounded_pop
            evaluation['score'] = v  # Value is used as score for front-end display
            
            total_city_population += rounded_pop
            total_developer_profit += rounded_pop * v
        else:
            state['value'] = 0
            state['population'] = 0
            evaluation['score'] = 0
            
    # 4. Third pass: Calculate total government tax
    tax_rate = 0.5
    total_government_tax = round(total_city_population * tax_rate)
    
    return {
        "faces": cloned_faces,
        "government_tax": total_government_tax,
        "developer_profit": total_developer_profit,
        "total_population": total_city_population
    }
